core/src/main/java/org/apache/calcite/sql/dialect/ClickHouseSqlDialect.java

```diff
@@ -118,36 +121,62 @@ public ClickHouseSqlDialect(Context context) {
-            String.format(Locale.ROOT, "FixedString(%s)",
-            type.getPrecision()), typeName, type.isNullable());
-      case VARCHAR:
-        return createSqlDataTypeSpecByName("String", typeName, type.isNullable());
-      case TINYINT:
-        return createSqlDataTypeSpecByName("Int8", typeName, type.isNullable());
-      case SMALLINT:
-        return createSqlDataTypeSpecByName("Int16", typeName, type.isNullable());
-      case INTEGER:
-        return createSqlDataTypeSpecByName("Int32", typeName, type.isNullable());
-      case BIGINT:
-        return createSqlDataTypeSpecByName("Int64", typeName, type.isNullable());
-      case REAL:
-        return createSqlDataTypeSpecByName("Float32", typeName, type.isNullable());
-      case FLOAT:
-      case DOUBLE:
-        return createSqlDataTypeSpecByName("Float64", typeName, type.isNullable());
-      case DATE:
-        return createSqlDataTypeSpecByName("Date", typeName, type.isNullable());
-      case TIMESTAMP:
-      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
-        return createSqlDataTypeSpecByName("DateTime", typeName, type.isNullable());
-      default:
-        break;
-      }
+    SqlTypeName typeName = type.getSqlTypeName();
+    switch (typeName) {
+    case CHAR:
+      return createSqlDataTypeSpecByName(
+          String.format(Locale.ROOT, "FixedString(%s)",
+              type.getPrecision()), typeName, type.isNullable());
+    case VARCHAR:
+      return createSqlDataTypeSpecByName("String", typeName, type.isNullable());
+    case TINYINT:
+      return createSqlDataTypeSpecByName("Int8", typeName, type.isNullable());
+    case SMALLINT:
+      return createSqlDataTypeSpecByName("Int16", typeName, type.isNullable());
+    case INTEGER:
+      return createSqlDataTypeSpecByName("Int32", typeName, type.isNullable());
+    case BIGINT:
+      return createSqlDataTypeSpecByName("Int64", typeName, type.isNullable());
+    case REAL:
+      return createSqlDataTypeSpecByName("Float32", typeName, type.isNullable());
+    case FLOAT:
+    case DOUBLE:
+      return createSqlDataTypeSpecByName("Float64", typeName, type.isNullable());
+    case DATE:
+      return createSqlDataTypeSpecByName("Date", typeName, type.isNullable());
+    case TIMESTAMP:
+    case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      return createSqlDataTypeSpecByName("DateTime", typeName, type.isNullable());
+    case MAP:
+      MapSqlType mapSqlType = (MapSqlType) type;
+      SqlDataTypeSpec keySpec = (SqlDataTypeSpec) getCastSpec(mapSqlType.getKeyType());
+      SqlDataTypeSpec valueSpec =
+          (SqlDataTypeSpec) getCastSpec(mapSqlType.getValueType());
+      SqlDataTypeSpec nonNullKeySpec =
+          requireNonNull(keySpec, "keySpec");
+      SqlDataTypeSpec nonNullValueSpec =
+          requireNonNull(valueSpec, "valueSpec");
+      SqlMapTypeNameSpec sqlMapTypeNameSpec =
+          new RelToSqlConverterUtil.ClickHouseSqlMapTypeNameSpec(nonNullKeySpec, nonNullValueSpec,
```

Comment:
I hope we can :
```
case MAP:
     return RelToSqlConverterUtil.getCastSpecClickHouseSqlMapType(xx,xxx,xx,);
case ARRAY:
     return RelToSqlConverterUtil.getCastSpecClickHouseSqlArrayType(xx,xxx,xx,);
````
then we add code comment in getCastSpecClickHouseSqlArrayType and getCastSpecClickHouseSqlMapType, to clarify what it was like before the transformation and what it is like after the transformation.